typescript/db/src/repository/index.ts

```diff
@@ -1,62 +1,115 @@
+import { prisma } from "./prisma-client";
 import {
   getUploadTargetHttp,
   getFromStorage,
   putInStorage,
   deleteFromStorage,
 } from "./upload-supabase";
-import { countLabels, listLabels } from "./label";
+import {
+  addWorkspace,
+  getWorkspace,
+  listWorkspace,
+  updateWorkspace,
+} from "./workspace";
+import { listLabels } from "./label";
 import { castObjectNullsToUndefined } from "./utils";
-
-export const prisma = new PrismaClient();
+import {
+  checkUserAccessToDataset,
+  checkUserAccessToImage,
+  checkUserAccessToLabel,
+  checkUserAccessToLabelClass,
+  checkUserAccessToWorkspace,
+} from "./access-control";
 
 export const repository: Repository = {
   image: {
-    add: async (image) => {
+    add: async (image, user) => {
+      await checkUserAccessToDataset({ where: { id: image.datasetId }, user });
       const createdImage = await prisma.image.create({
         data: castObjectNullsToUndefined(image),
       });
       return createdImage.id;
     },
-    count: (where) =>
-      prisma.image.count({
-        where: castObjectNullsToUndefined(where),
-      }),
-    get: async (where) =>
-      (await prisma.image.findUnique({
+    count: async (whereWithUser) => {
+      const { user, ...where } = whereWithUser ?? { user: undefined };
+      return await prisma.image.count({
+        where: castObjectNullsToUndefined({
+          ...where,
+          dataset: {
+            workspace: { memberships: { some: { userId: user?.id } } },
+          },
+        }),
+      });
+    },
+    get: async (where, user) => {
+      await checkUserAccessToImage({ where, user });
+      return (await prisma.image.findUnique({
         where,
-      })) as unknown as Image,
+      })) as unknown as Image;
+    },
 
-    list: (where, skip = undefined, first = undefined) =>
-      prisma.image.findMany(
+    list: async (whereWithUser, skip = undefined, first = undefined) => {
+      const { user, ...where } = whereWithUser ?? { user: undefined };
+      return await prisma.image.findMany(
         castObjectNullsToUndefined({
-          where: castObjectNullsToUndefined(where),
+          where: castObjectNullsToUndefined({
+            ...where,
+            dataset: {
+              workspace: { memberships: { some: { userId: user?.id } } },
+            },
+          }),
           orderBy: { createdAt: Prisma.SortOrder.asc },
           skip,
           take: first,
         })
-      ),
+      );
+    },
   },
   label: {
-    add: async (label) => {
+    add: async (label, user) => {
+      await checkUserAccessToImage({ where: { id: label.imageId }, user });
       const createdLabel = await prisma.label.create({ data: label });
       return createdLabel.id;
     },
-    count: countLabels,
-    delete: async ({ id }) => {
+    count: async (whereWithUser) => {
+      if (!whereWithUser) {
+        return 0;
+      }
+      const { user, ...where } = whereWithUser;
+      if ("datasetId" in where) {
```

Comment:
Shouldn't we add something like `checkUserAccessToDataset` here?

Reply:
Yes you're totally right, I'll add it, nice catch!

Reply:
Added in 928c65a8cc5638b940bd2f6023aa02b3b0caac42. Actually it shouldn't be needed because the only way we can list or count labels of a dataset is through a dataset query where we already test if the user has access to it. But I guess it does not hurt to add this